compiler/qsc_parse/src/lex/formatter.rs

```diff
@@ -0,0 +1,270 @@
+                }
+
+                if let ConcreteTokenKind::WhiteSpace = one.kind {
+                    // first token is whitespace, continue scanning
+                    continue;
+                } else if let ConcreteTokenKind::WhiteSpace = two.kind {
+                    // whitespace in the middle
+                    apply_rules(
+                        one,
+                        get_token_contents(code, two),
+                        three,
+                        code,
+                        indent_level,
+                    )
+                } else {
+                    // one, two are adjacent tokens with no whitespace in the middle
+                    apply_rules(one, "", two, code, indent_level)
+                }
+            }
+            (None, None, Some(three)) => {
+                // Remove any whitespace at the start of a file
+                if three.span.lo != 0 {
+                    vec![Edit::new(0, three.span.lo, "")]
+                } else {
+                    vec![]
+                }
+            }
+            _ => {
+                // not enough tokens to apply a rule
+                // TODO: we'll probably need to handle end-of-file cases here
+                continue;
+            }
+        };
+
+        edits.append(&mut edits_for_triple);
+    }
+
+    edits
+}
+
+fn fix_whitespace(whitespace: &str, indent_level: usize) -> String {
+    //
+    // when you see newline, insert the indent string
+    // and trim until the next newline or the end of the string
+    //
+
+    let mut count_newlines = whitespace.chars().filter(|c| *c == '\n').count();
+
+    // There should always be at least one newline
+    if count_newlines < 1 {
+        count_newlines = 1;
+    }
+    let mut new = "\n".repeat(count_newlines);
+    new.push_str(&make_indent_string(indent_level));
+    new
+}
+
+fn apply_rules(
+    left: &ConcreteToken,
+    whitespace: &str,
+    right: &ConcreteToken,
+    code: &str,
+    indent_level: usize,
+) -> Vec<Edit> {
+    let mut edits = vec![];
+    // when we get here, neither left nor right should be whitespace
+
+    // if the right is a close brace, the indent level should be one less
+    let indent_level = if let ConcreteTokenKind::Cooked(TokenKind::Close(Delim::Brace)) = right.kind
+    {
+        if indent_level > 0 {
+            indent_level - 1
+        } else {
+            indent_level
+        }
+    } else {
+        indent_level
+    };
+
+    match (&left.kind, &right.kind) {
+        (
+            ConcreteTokenKind::Cooked(TokenKind::Open(l)),
+            ConcreteTokenKind::Cooked(TokenKind::Close(r)),
+        ) if l == r => {
+            rule_no_space(left, whitespace, right, &mut edits);
+        }
+        (ConcreteTokenKind::Comment | ConcreteTokenKind::Cooked(TokenKind::DocComment), _) => {
+            rule_trim_comments(left, &mut edits, code);
+            rule_indentation(left, whitespace, right, &mut edits, indent_level);
```

Comment:
I think, in theory, you don't need to have the `rule_indentation` calls sprinkled throughout the `match` expression like this. If the `rule_` functions returned the _new whitespace_ instead of the edits, you could rewrite this to look like:

1. Big match statement - apply `rule_xyz` functions to get `new_whitespace`
2. _After_ the match statement, call `fix_whitespace` to update `new_whitespace`
3. If new whitespace is not equal 
4. If `new_whitespace != old_whitespace` , generate the edits 

This would be more readable (and easier to follow logically) I think.

Reply:
Not all of the rules rewrite the whitespace. Look at the `rule_trim_comments` rule, which edits the contents of the comment.